### DISCUSSION TITLE
Can it be used without the [locale] folder?

### DISCUSSION THREAD

**Discussion by sawa-ko** [amannn/next-intl#1016] [Q&A]
Actually I want to use this library but I don't want to use the [locale] folder, I don't know if there is a cookie based way or something like that.

I hope you can help me.

**Comment by amannn** [comment]
Please see https://github.com/amannn/next-intl/issues/960!

**Reply by amannn** [reply]
@sawa-ko As an update, there's now a proof-of-concept available for this feature: https://github.com/amannn/next-intl/pull/1017.

If you'd like to take this for a test drive, feedback is very much appreciated!

**Reply by sawa-ko** [reply]
Cool, it looks good, this way you can also get that language through the cookies function of next or what would it be like? This because server/client components

**Reply by amannn** [reply]
Yep, you can read from `cookies`, other headers or a user preference if you store such a setting in your database!

**Reply by sawa-ko** [reply]
I got this error:

```
Error: failed to pipe response at et (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:386745) at runNextTicks (node:internal/process/task_queues:60:5) at process.processImmediate (node:internal/timers:449:9) at process.callbackTrampoline (node:internal/async_hooks:130:17) at async er.pipeToNodeResponse (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:388194) at async sendRenderResult (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\send-payload.js:118:5) at async DevServer.pipeImpl (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\base-server.js:935:13) at async NextNodeServer.handleCatchallRenderRequest (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\next-server.js:272:17) at async DevServer.handleRequestImpl (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\base-server.js:816:17) at async C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\dev\next-dev-server.js:339:20 at async Span.traceAsyncFn (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\trace\trace.js:154:20) at async DevServer.handleRequest (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\dev\next-dev-server.js:336:24) at async invokeRender (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\lib\router-server.js:174:21) at async handleRequest (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\lib\router-server.js:353:24) at async requestHandlerImpl (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\lib\router-server.js:377:13) at async Server.requestListener (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\server\lib\start-server.js:141:13) {stack: "Error: failed to pipe response\n at et (C:\Users…ules\next\dist\server\lib\start-server.js:141:13)", message: "failed to pipe response", cause: Error: Expected a suspended thenable. This is a bug in React. Please file an issue.\n at B (C:\Us…}
```

```
Internal error: TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed at ReadableByteStreamController.enqueue (node:internal/webstreams/readablestream:1179:13) at eV (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:276025) at eJ (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:275323) at Timeout._onTimeout (C:\Users\---\WebstormProjects\logistic-system\node_modules\.pnpm\next@14.2.2_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0__react@18.2.0\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:35:265080) at listOnTimeout (node:internal/timers:573:17) at process.processTimers (node:internal/timers:514:7)
```

```tsx
// layout.tsx - server component

export default async function RootLayout({ children, params }: WithPagePropsAndChildren) {
  const locale = useLocale();
  const messages = await getMessages({ locale: params.locale });

  return (
    <html lang={params.locale} suppressHydrationWarning={true}>
      <body className={twClass(inter.variable, 'min-h-screen bg-background font-sans antialiased')}>
        <Providers locale={locale} localeMessages={messages}>
          {children}
        </Providers>
      </body>
    </html>
  );
}

// providers.tsx - client component
export default function Providers({ children, locale, localeMessages }: ProvidersProps) {
  return (
    <ThemeProvider attribute="class" defaultTheme="system" enableColorScheme={true} enableSystem={true} storageKey={'__APP_THEME__'}>
      <NextIntlClientProvider locale={locale} messages={localeMessages}>
        {children}
      </NextIntlClientProvider>
    </ThemeProvider>
  );
}
```

```ts
// i18n.ts

export default getRequestConfig(async ({ locale }) => {
  if (!LOCALES.includes(locale as any)) notFound();

  return {
    locale: await getCookieLocale(),
    messages: (await import(`./locales/${locale}.json`)).default,
  };
});

export default async function getCookieLocale() {
  const cookieStore = cookies();
  return cookieStore.get(LOCALE_COOKIE)?.value || LOCALES[0];
}
```

**Reply by amannn** [reply]
There are some issues here:
1. If your `RootLayout` is `async`, you can't call a hook like `useLocale`. Instead, use `await getLocale()` (see [Using internationalization in Server Components](https://next-intl-docs.vercel.app/docs/environments/server-client-components#using-internationalization-in-server-components)).
2. Does your `RootLayout` still accept `params`? The feature is intended to be used without a `[locale]` folder, hence there should be no `params.locale`.
3. You shouldn't read `async ({ locale })` in `getRequestConfig`. Instead, simply return the locale you're reading via `getCookieLocale`.

Have a look at the working implementation that's linked in the pull request when in doubt! Hope this helps! 🙂

**Reply by sawa-ko** [reply]
Now this 😆 

![image](https://github.com/amannn/next-intl/assets/56084970/1cd8bbd1-3ddf-4a11-a8aa-28f47bb86121)

![image](https://github.com/amannn/next-intl/assets/56084970/9f18868d-7c6d-4945-94a8-8a0b672565dd)

> I am not using it anywhere

![image](https://github.com/amannn/next-intl/assets/56084970/3744d220-e225-4ba4-8808-61f005f5ecac)

**Reply by amannn** [reply]
Can you check if there’s an error on your server console? Anything you’re doing differently than in the example I’ve linked to?

**Reply by sawa-ko** [reply]
No errors! Just this overlay error!

**Reply by amannn** [reply]
Is that during development or production? next-intl includes additional warnings during development that might be helpful. Are you by chance using the navigation APIs?

Currently, this feature is intended to be used with the following assumptions (as mentioned in the PR):

1. No `[locale]` segment needs to be added
2. No middleware needs to be configured
3. The navigation APIs from `next-intl` don't need to be used
4. Static rendering works without tradeoffs

I'm having a look if I can support the navigation APIs nonetheless or at least provide better error handling.

**Reply by sawa-ko** [reply]
Hum, I have them declared but I don't have them using, and it is in development that the error message is displayed.

To be honest, I haven't noticed if there are any warnings on the console, I'll check.

**Reply by sawa-ko** [reply]
Yes, it seems that I had it imported in a component and I have removed it and it seems that it has been solved. A bit weird, not sure if that was the problem.

May I make a suggestion or ask a question? Why is the typing not working here or is it not typed?

![image](https://github.com/amannn/next-intl/assets/56084970/24eaab05-e923-464c-a6d9-9d8376792899)

**Reply by amannn** [reply]
Great to hear if that fixed the issue for you! 🙌 

Hmm, interesting that you don't see any completions there, theoretically it should work:

![Screenshot 2024-04-25 at 16 29 31](https://github.com/amannn/next-intl/assets/4038316/0ce5a471-e54e-4bb1-a428-077c956e6929)

Note however that `locale` is not included in the types yet for the release you're using, I noticed that after the pre-release and have fixed it in https://github.com/amannn/next-intl/pull/1017/commits/2375929361baa2c7a7c78e2f327c6445eaacbf36.

**Reply by sawa-ko** [reply]
Great, and it seems to be working fine, I've been testing it for a couple of hours and haven't encountered any message problems. 💯
